Microsoft.Toolkit.Uwp.UI.Controls/StaggeredPanel/StaggeredPanel.cs

```diff
@@ -0,0 +1,155 @@
+            if (HorizontalAlignment == HorizontalAlignment.Stretch)
+            {
+                _columnWidth = availableSize.Width / numColumns;
+            }
+
+            var columnHeights = new double[numColumns];
+
+            for (int i = 0; i < Children.Count; i++)
+            {
+                var columnIndex = GetColumnIndex(columnHeights);
+
+                var child = Children[i];
+                child.Measure(new Size(_columnWidth, availableSize.Height));
+                var elementSize = child.DesiredSize;
+                columnHeights[columnIndex] += elementSize.Height;
+            }
+
+            double desiredHeight = columnHeights.Max();
+
+            return new Size(availableSize.Width, desiredHeight);
+        }
+
+        /// <inheritdoc/>
+        protected override Size ArrangeOverride(Size finalSize)
+        {
+            double offset = 0;
+            int numColumns = (int)Math.Floor(finalSize.Width / _columnWidth);
+            if (HorizontalAlignment == HorizontalAlignment.Right)
+            {
+                offset = finalSize.Width - (numColumns * _columnWidth);
+            }
+            else if (HorizontalAlignment == HorizontalAlignment.Center)
+            {
+                offset = (finalSize.Width - (numColumns * _columnWidth)) / 2;
+            }
+
+            var columnHeights = new double[numColumns];
+
+            for (int i = 0; i < Children.Count; i++)
+            {
+                var columnIndex = GetColumnIndex(columnHeights);
+
+                var child = Children[i];
+                var elementSize = child.DesiredSize;
+
+                double elementWidth = elementSize.Width;
+                double elementHeight = elementSize.Height;
+                if (elementWidth > _columnWidth)
+                {
+                    double differencePercentage = _columnWidth / elementWidth;
+                    elementHeight = elementHeight * differencePercentage;
+                    elementWidth = _columnWidth;
+                }
+
+                Rect bounds = new Rect(offset + (_columnWidth * columnIndex), columnHeights[columnIndex], elementWidth, elementHeight);
+                child.Arrange(bounds);
+
+                columnHeights[columnIndex] += elementSize.Height;
+            }
+
+            return base.ArrangeOverride(finalSize);
+        }
+
+        private static void OnDesiredColumnWidthChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if (d is StaggeredPanel panel)
+            {
+                panel.InvalidateMeasure();
+            }
+        }
+
+        private void OnHorizontalAlignmentChanged(DependencyObject sender, DependencyProperty dp)
+        {
+            InvalidateMeasure();
+        }
+
+        private int GetColumnIndex(double[] columnHeights)
+        {
+            int columnIndex = 0;
+            double height = columnHeights[0];
```

Comment:
Got an exeption here when the width of the control was less than the first column.

![image](https://user-images.githubusercontent.com/711864/35128351-018ed788-fc6b-11e7-81c7-914d2f876c27.png)

Reply:
Good catch! fixed.